File: leetcode-js/091-decode-ways.js
/**
 * @param {string} s
 * @return {number}
 */
var numDecodings = function(s) {
    if (s.length === 0) {
        return 0;
    }

    const dp = Array(s.length + 1).fill(0);

    dp[s.length] = 1;

    if (s[s.length -1] !== '0') {
        dp[s.length-1] = 1;
    }

    for (let i = s.length - 2; i >= 0; i -= 1) {
        if (s.charAt(i) !== '0') {
            dp[i] = parseInt(s.substr(i, 2)) <= 26 ? dp[i+1] + dp[i+2] : dp[i+1];
        }
    }

    return dp[0];
};

File: leetcode-java/091-decode-ways.java
public class Solution {
    public int numDecodings(String s) {
        int l = s.length();

        if (l == 0) {
            return 0;
        }

        int second = 1;
        int first = s.charAt(l - 1) == '0' ? 0 : 1;

        for (int i = l - 2; i >= 0; i--) {
            if (s.charAt(i) == '0') {
                second = first;
                first = 0;
            } else if (i < l - 1 && Integer.parseInt(s.substring(i, i + 2)) <= 26) {
                // don't have to parseInt, can save int value from last time

                int t = first;
                first = first + second;
                second = t;
            } else {
                second = first;
            }
        }

        return first;
    }
}
